doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1083,45 +1403,52 @@ Testing boolean expressions
    .. exn:: Condition not satisfied.
       :undocumented:
 
-Proving a subgoal as a separate lemma
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Proving a subgoal as a separate lemma: abstract
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. todo PR should this be adjacent to the description of "solve"?
+   Should the reference below to :n:`solve @ltac_expr` be to solve [ ... ]?
 
-.. tacn:: abstract @ltac_expr
-   :name: abstract
+.. tacn:: abstract @ltac_expr2 {? using @ident }
 
-   From the outside, :n:`abstract @ltac_expr` is the same as :n:`solve @ltac_expr`.
-   Internally it saves an auxiliary lemma called ``ident_subproofn`` where
-   ``ident`` is the name of the current goal and ``n`` is chosen so that this is
-   a fresh name. Such an auxiliary lemma is inlined in the final proof term.
+   Does a :tacn:`solve` :n:`@ltac_expr` and saves the subproof as an auxiliary lemma.
+   if :token:`ident` is specified, the lemma is saved with that name; otherwise
+   the lemma is saved with the name `NAME_subproofN` where
+   `NAME` is the name of the current goal and `N` is
+   a fresh name. The auxiliary lemma is inlined in the final proof term.
 
-   This tactical is useful with tactics such as :tacn:`omega` or
-   :tacn:`discriminate` that generate huge proof terms. With that tool the user
-   can avoid the explosion at time of the Save command without having to cut
-   manually the proof in smaller lemmas.
+   This is useful with tactics such as :tacn:`omega` or
+   :tacn:`discriminate` that generate huge proof terms. It makes the
+   the subsequent Save command faster without having to
+   manually split the proof into smaller lemmas.
+
+   .. todo PR Does this mean a net time saving across the entire proof
+      or is it just moving some of the Save work into abstract?  Or is
+      some non-linear, non-zero sum effect here?
```

Comment:
I believe there are rare corner cases where evar normalization is non-linear in the number of evars, and so abstracting can result in a net time saving.  But this is pretty rare, I believe, and most of the time `abstract` makes the proof a tiny bit slower.  However, there is an enormous saving of peak memory usage, because `abstract` will compact the proof state early, resulting in the ability to reuse memory across branches.  Additionally, you can decrease the *net* time of reflective proofs by up to a factor of 2 by using the pattern `abstract exact_no_check foo`, wherein you fail fast when the `exact` fails, but you don't need to check that the exact succeeds twice.

Reply:
Just curious, "compacting the proof state" meaning discarding proof steps/states within Stm that are associated with the `abstract`?  I'd love to understand what factors drive peak memory usage--does that mostly affect only complex/long-running Ltac tactics?  At first thought, I'd expect memory usage would grow as the product of the proof term and the number of proof steps.

Reply:
How about this description?

   This is useful with tactics such as :tacn:`omega` or
   :tacn:`discriminate` that generate huge proof terms.  It can
   significantly reduce peak memory use.  In most cases it doesn't
   have a significant impact on run time.  One case in which it can
   help is when a tactic `foo` is known to always pass type checking when it
   succeeds, such as in reflective proofs.  In this case, the idiom
   ":tacn:`abstract` :tacn:`exact_no_check` `foo`" will save half the type
   checking type time compared to ":tacn:`exact` `foo`".

Reply:
> Just curious, "compacting the proof state" meaning discarding proof steps/states within Stm that are associated with the `abstract`? I'd love to understand what factors drive peak memory usage--does that mostly affect only complex/long-running Ltac tactics? At first thought, I'd expect memory usage would grow as the product of the proof term and the number of proof steps.

I don't believe it's the stm, since the stm registers no states, AIUI, within a single sentence.  I believe instead it is about evar normalization and discarding the evar representation of instantiated evars.  I believe the relevant operation is the one invoked by the vernacular `Optimize Proof`.  I don't have a good sense of memory usage, in large part due to lack of fine-grained profiling, cf https://github.com/coq/coq/issues/11926 and https://github.com/coq/coq/issues/11926

Reply:
> How about this description?
> 
> This is useful with tactics such as :tacn:`omega` or
> :tacn:`discriminate` that generate huge proof terms. It can
> significantly reduce peak memory use. In most cases it doesn't
> have a significant impact on run time. One case in which it can
> help is when a tactic `foo` is known to always pass type checking when it
> succeeds, such as in reflective proofs. In this case, the idiom
> ":tacn:`abstract` :tacn:`exact_no_check` `foo`" will save half the type
> checking type time compared to ":tacn:`exact` `foo`".

I'd suggest "generate huge proof terms with many intermediate goals" or "generate huge proof terms with many intermediate evars" (or perhaps "internal" rather than "intermediate") rather than just "generate huge proof terms".  I don't think it's just about the size of the proof term, but about how expensive it is to evar-normalize the proof term.

Maybe "One case in which it can help with performance" instead of "One case in which it can help".  Helping with memory is already helping.  (Or alternatively, use "One other" or "One other notable" instead of just "One" to start this sentence.)

Otherwise, that description looks good to me!